relays/lib-substrate-relay/src/messages_source.rs

```diff
@@ -199,11 +201,97 @@ where
 			)
 			.await?;
 
-		make_message_details_map::<P::SourceChain>(
+		let mut messages = make_message_details_map::<P::SourceChain>(
 			Decode::decode(&mut &encoded_response.0[..])
 				.map_err(SubstrateError::ResponseParseFailed)?,
 			nonces,
-		)
+		)?;
+
+		// prepare arguments of the inbound message details call (if we need it)
+		let mut messages_to_refine = HashMap::new();
+		for (message_nonce, message) in &messages {
+			if message.dispatch_fee_payment != DispatchFeePayment::AtTargetChain {
+				continue
+			}
+
+			// for pay-at-target messages we may want to ask target chain for
+			// refined dispatch weight
+			let message_key = bp_messages::storage_keys::message_key(
+				P::TargetChain::WITH_CHAIN_MESSAGES_PALLET_NAME,
+				&self.lane_id,
+				*message_nonce,
+			);
+			let message_data: MessageData<BalanceOf<P::SourceChain>> =
+				self.source_client.storage_value(message_key, Some(id.1)).await?.ok_or_else(
+					|| {
+						SubstrateError::Custom(format!(
+							"Message to {} {:?}/{} is missing from runtime the storage of {} at {:?}",
+							P::TargetChain::NAME,
+							self.lane_id,
+							message_nonce,
+							P::SourceChain::NAME,
+							id,
+						))
+					},
+				)?;
+			let message_payload = message_data.payload;
```

Comment:
I wouldn't rather touch this enum, since e.g. in encoded-call messaging it is stored as a part of message payload. The alternative would be to add message payload to the `message_details` method result. But messages may be large and probably can cause a oom in WASM, if we're calling `message_details` for some range that includes multiple large messages (and we can't know if it'll oom in advance). So probably better to leave it as is (imo)

Reply:
Sure, it's better to keep it as it is if it's safer. But just one more question:

> But messages may be large and probably can cause a oom in WASM, if we're calling message_details for some range that includes multiple large messages (and we can't know if it'll oom in advance)

Than is it possible for WASM to oom when calling the following as well ?
```
			let refined_messages_encoded = self
				.target_client
				.state_call(
					P::SourceChain::FROM_CHAIN_MESSAGE_DETAILS_METHOD.into(),
					Bytes((self.lane_id, messages_to_refine.values().collect::<Vec<_>>()).encode()),
					None,
				)
				.await?;
```
Since `messages_to_refine` contains the payloads.

Reply:
Yeah, that's a valid concern. Then it worth to change `FROM_CHAIN_MESSAGE_DETAILS_METHOD` runtime API method so that it'll accept single message only. Could you, please, file an issue for that?

Reply:
Sure. Also I'd be happy to assign it to me if that's ok for you.

Reply:
Sure, go on! :) Thank you!